polars/polars-arrow/src/kernels/rolling/nulls/min_max.rs

```diff
@@ -1,460 +1,247 @@
-use arrow::bitmap::utils::{count_zeros, ZipValidityIter};
 use nulls;
 use nulls::{rolling_apply_agg_window, RollingAggWindowNulls};
 
 use super::*;
 
-pub fn is_reverse_sorted_max_nulls<T: NativeType + PartialOrd + IsFloat>(
-    values: &[T],
-    validity: &Bitmap,
-) -> bool {
-    let mut current_max = None;
-    for opt_v in ZipValidityIter::new(values.iter(), validity.iter()) {
-        match (current_max, opt_v) {
-            // do nothing
-            (None, None) => {}
-            (None, Some(v)) => current_max = Some(*v),
-            (Some(current), Some(val)) => {
-                match compare_fn_nan_min(&current, val) {
-                    Ordering::Greater => {
-                        current_max = Some(*val);
-                    }
-                    // allowed
-                    Ordering::Equal => {}
-                    // not sorted
-                    Ordering::Less => return false,
-                }
-            }
-            (Some(_current), None) => {}
-        }
-    }
-
-    true
+#[inline]
+fn new_is_min<T: NativeType + IsFloat + PartialOrd>(old: &T, new: &T) -> bool {
+    compare_fn_nan_min(old, new).is_ge()
 }
 
-pub struct SortedMinMax<'a, T: NativeType> {
-    slice: &'a [T],
-    validity: &'a Bitmap,
-    last_start: usize,
-    last_end: usize,
-    null_count: usize,
+#[inline]
+fn new_is_max<T: NativeType + IsFloat + PartialOrd>(old: &T, new: &T) -> bool {
+    compare_fn_nan_max(old, new).is_le()
 }
 
-impl<'a, T: NativeType> SortedMinMax<'a, T> {
-    fn count_nulls(&self, start: usize, end: usize) -> usize {
-        let (bytes, offset, _) = self.validity.as_slice();
-        count_zeros(bytes, offset + start, end - start)
-    }
-}
-
-impl<'a, T: NativeType> RollingAggWindowNulls<'a, T> for SortedMinMax<'a, T> {
-    unsafe fn new(
-        slice: &'a [T],
-        validity: &'a Bitmap,
-        start: usize,
-        end: usize,
-        _params: DynArgs,
-    ) -> Self {
-        let mut out = Self {
-            slice,
-            validity,
-            last_start: start,
-            last_end: end,
-            null_count: 0,
-        };
-        let nulls = out.count_nulls(start, end);
-        out.null_count = nulls;
-        out
-    }
-
-    unsafe fn update(&mut self, start: usize, end: usize) -> Option<T> {
-        self.null_count -= self.count_nulls(self.last_start, start);
-        self.null_count += self.count_nulls(self.last_end, end);
-
-        self.last_start = start;
-        self.last_end = end;
-
-        // return first non null
-        for idx in start..end {
-            let valid = self.validity.get_bit_unchecked(idx);
-
-            if valid {
-                return Some(*self.slice.get_unchecked(idx));
+// Identify runs of sorted values, either increasing or decreasing. It lets us avoid unnecessary
+// comparisons when the extremum keeps dropping off the front of the window
+macro_rules! fn_n_sorted_past {
```

Comment:
Can we rewrite this macro with generics? I only want macro's to do dispatch and very small code things like `polars_err!`. 

The parts you want variable, you can put behind auxilliary traits. Happy to help if needed.

Reply:
Ok, now it's slower. Even after I changed the code a bit to try and optimize it's better than my first attempt with  generics but still slower than the macro version. I'll take another crack at it but maybe I'm missing something obvious.

Reply:
A little better with some const generic hackery. Tolerable now but still looks a little slower on average than the macro version.

Reply:
Any thoughts? I can make it a tiny bit faster by making the compare part branchless now but I don't think it matters. Maybe there are other improvements too but it seems alright. Are you happy with this style overall?

Reply:
What else do we need to do to get this landed. It fixes some pretty nasty edge cases and I'd like to put rolling min/max behind me for a while :-)

Reply:
Didn't get to it yet. :) Its on my stack!

Reply:
Ok, no problem. I know big-ish ones like this probably take more effort to review.